README.md
# Reddit_DataMine
The Data Mining Program for the <a href="https://github.com/IceCereal/Bot_REGOD-New">Bot_REGOD-new</a> which collected <a href="https://github.com/IceCereal/Bot_REGOD-New/tree/master/Modules/SubredditData">Reddit Data</a> for one day. The program tries to look for patterns in the collected data.

Modules:<br/><br/>
`readSubredditData.py` - Data Preprocessing
```python
from readSubredditData import Reader

subreddit = Reader('todayilearned')
(X, y) = subreddit.getData(type)
```
`type` (`str`): <br/>
`'high'`  :   returns the highest score in the list (<b>DEFAULT</b>)<br/>
`'mean'`  :   returns the mean of the values in the list<br/>
`'sigma'` :   returns the sum of all the values of the list<br/>


SubredditData/news
$['2018-11-14 21:24:07.666628', [9, 7, 5, 5, 4, 3, 2, 1, 0, 0]]$['2018-11-14 21:40:06.371701', [7, 4, 0, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-14 21:55:59.586074', [10, 4, 4, 4, 2, 1, 1, 0, 0, 0]]$['2018-11-14 22:11:53.943538', [16, 6, 6, 5, 5, 4, 4, 4, 3, 3]]$['2018-11-14 22:27:51.278823', [12, 8, 7, 3, 0, 0, 0, 0, 0, 0]]$['2018-11-14 22:43:45.102857', [22, 13, 13, 11, 5, 4, 0, 0, 0, 0]]$['2018-11-14 22:59:38.990097', [25, 23, 5, 3, 1, 0, 0, 0, 0, 0]]$['2018-11-14 23:15:31.428767', [48, 43, 10, 8, 4, 4, 2, 1, 0, 0]]$['2018-11-14 23:31:26.960647', [258, 64, 12, 8, 6, 6, 5, 5, 3, 3]]$['2018-11-14 23:47:18.114805', [564, 16, 4, 3, 2, 0, 0, 0, 0, 0]]$['2018-11-15 00:03:09.654980', [955, 10, 8, 8, 8, 6, 5, 4, 0, 0]]$['2018-11-15 00:19:01.641147', [22, 22, 16, 14, 13, 10, 8, 0, 0, 0]]$['2018-11-15 00:34:55.964674', [37, 26, 25, 23, 10, 7, 6, 0, 0, 0]]$['2018-11-15 00:50:47.072724', [74, 36, 14, 12, 11, 10, 0, 0, 0, 0]]$['2018-11-15 01:06:36.769227', [23, 22, 11, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-15 01:22:26.902453', [28, 28, 12, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-15 01:38:16.772857', [16, 1, 0, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-15 01:54:05.718659', [11, 9, 5, 4, 2, 2, 1, 0, 0, 0]]$['2018-11-15 02:09:55.610217', [18, 18, 9, 6, 5, 3, 3, 1, 0, 0]]$['2018-11-15 02:25:52.811319', [32, 25, 17, 8, 8, 8, 7, 5, 5, 4]]$['2018-11-15 02:41:42.613992', [33, 21, 12, 11, 11, 9, 9, 5, 5, 5]]$['2018-11-15 02:57:39.817839', [19, 17, 15, 11, 9, 7, 6, 5, 5, 4]]$['2018-11-15 03:13:29.183246', [29, 19, 19, 8, 8, 7, 7, 7, 6, 3]]$['2018-11-15 03:29:21.264302', [18, 10, 8, 5, 4, 0, 0, 0, 0, 0]]$['2018-11-15 03:45:13.308566', [28, 10, 9, 4, 0, 0, 0, 0, 0, 0]]$['2018-11-15 04:01:01.467649', [17, 12, 0, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-15 04:16:51.205154', [30, 9, 5, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-15 04:32:42.903046', [50, 19, 13, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-15 04:48:32.144826', [25, 20, 2, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-15 05:04:19.457038', [33, 5, 4, 4, 3, 2, 1, 0, 0, 0]]$['2018-11-15 05:20:07.031277', [10, 9, 8, 7, 7, 4, 2, 0, 0, 0]]$['2018-11-15 05:35:51.043050', [22, 11, 11, 8, 7, 4, 0, 0, 0, 0]]$['2018-11-15 05:51:37.659871', [34, 18, 8, 7, 7, 2, 0, 0, 0, 0]]$['2018-11-15 06:07:20.815936', [44, 11, 6, 3, 1, 0, 0, 0, 0, 0]]$['2018-11-15 06:23:04.451623', [6, 6, 3, 2, 1, 0, 0, 0, 0, 0]]$['2018-11-15 06:38:51.040331', [17, 15, 3, 1, 0, 0, 0, 0, 0, 0]]$['2018-11-15 06:54:34.897918', [26, 25, 3, 2, 1, 0, 0, 0, 0, 0]]$['2018-11-15 07:10:18.031078', [32, 30, 8, 2, 2, 1, 0, 0, 0, 0]]$['2018-11-15 07:26:00.898943', [8, 4, 3, 2, 2, 0, 0, 0, 0, 0]]$['2018-11-15 07:41:45.000123', [14, 8, 3, 3, 2, 1, 0, 0, 0, 0]]$['2018-11-15 07:57:26.569338', [14, 4, 3, 3, 2, 0, 0, 0, 0, 0]]$['2018-11-15 08:13:06.472728', [14, 6, 5, 4, 0, 0, 0, 0, 0, 0]]$['2018-11-15 08:28:49.820803', [11, 8, 4, 2, 1, 0, 0, 0, 0, 0]]$['2018-11-15 08:44:31.877649', [13, 7, 4, 3, 0, 0, 0, 0, 0, 0]]$['2018-11-15 09:00:15.351889', [17, 8, 7, 5, 0, 0, 0, 0, 0, 0]]$['2018-11-15 09:15:54.926454', [14, 9, 5, 4, 3, 2, 0, 0, 0, 0]]$['2018-11-15 09:31:38.274091', [6, 5, 4, 3, 2, 0, 0, 0, 0, 0]]$['2018-11-15 09:47:21.897504', [7, 7, 5, 4, 3, 0, 0, 0, 0, 0]]$['2018-11-15 10:03:05.478090', [7, 5, 4, 3, 0, 0, 0, 0, 0, 0]]$['2018-11-15 10:18:50.644667', [2, 1, 0, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-15 10:34:30.208413', [6, 1, 1, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-15 10:50:10.218111', [9, 2, 1, 0, 0, 0, 0, 0, 0, 0]]$['2018-11-15 11:05:51.206445', [12, 4, 4, 3, 1, 0, 0, 0, 0, 0]]$['2018-11-15 11:21:31.999094', [16, 8, 5, 5, 5, 5, 0, 0, 0, 0]]$['2018-11-15 11:37:12.718374', [10, 10, 8, 7, 4, 4, 0, 0, 0, 0]]$['2018-11-15 11:52:53.780177', [22, 15, 11, 11, 6, 5, 3, 2, 0, 0]]$['2018-11-15 12:08:38.255120', [27, 8, 6, 4, 3, 2, 2, 0, 0, 0]]$['2018-11-15 12:24:20.474310', [10, 10, 8, 4, 4, 3, 3, 3, 0, 0]]$['2018-11-15 12:40:03.521830', [15, 14, 6, 6, 4, 4, 3, 1, 0, 0]]$['2018-11-15 12:55:47.079201', [10, 6, 5, 3, 3, 1, 0, 0, 0, 0]]$['2018-11-15 13:11:39.100460', [11, 6, 6, 3, 3, 1, 0, 0, 0, 0]]$['2018-11-15 13:27:24.396261', [9, 7, 7, 6, 5, 4, 0, 0, 0, 0]]$['2018-11-15 13:43:09.925192', [17, 11, 7, 6, 6, 5, 0, 0, 0, 0]]$['2018-11-15 13:58:58.360958', [31, 8, 7, 6, 5, 2, 2, 1, 0, 0]]$['2018-11-15 15:41:42.736177', [28, 5, 4, 2, 1, 0, 0, 0, 0, 0]]$['2018-11-15 15:57:32.818191', [43, 8, 7, 6, 2, 2, 1, 0, 0, 0]]$['2018-11-15 16:13:22.016146', [13, 10, 6, 3, 1, 1, 0, 0, 0, 0]]$['2018-11-15 16:29:14.890604', [14, 13, 12, 8, 4, 3, 2, 2, 1, 0]]$['2018-11-15 16:50:33.486953', [21, 21, 10, 9, 8, 6, 6, 5, 4, 3]]$['2018-11-15 17:06:32.393628', [30, 24, 18, 17, 13, 12, 11, 9, 9, 7]]$['2018-11-15 17:22:26.091942', [36, 34, 24, 14, 11, 10, 4, 2, 1, 1]]$['2018-11-15 17:38:20.207970', [70, 62, 46, 18, 17, 13, 7, 5, 4, 3]]$['2018-11-15 17:54:16.222888', [92, 18, 11, 7, 7, 5, 5, 4, 2, 2]]$['2018-11-15 18:10:09.812278', [20, 12, 10, 8, 7, 6, 6, 6, 4, 3]]$['2018-11-15 18:26:05.501415', [26, 11, 7, 2, 0, 0, 0, 0, 0, 0]]$['2018-11-15 18:42:02.721186', [16, 9, 5, 3, 2, 1, 0, 0, 0, 0]]$['2018-11-15 18:57:58.291444', [40, 12, 9, 5, 4, 3, 3, 0, 0, 0]]$['2018-11-15 19:13:51.371712', [95, 24, 13, 10, 8, 7, 5, 4, 3, 2]]$['2018-11-15 19:29:47.533984', [196, 37, 23, 12, 8, 5, 2, 0, 0, 0]]$['2018-11-15 19:45:41.771530', [31, 15, 15, 11, 7, 4, 4, 1, 1, 0]]$['2018-11-15 20:01:36.031841', [25, 22, 14, 11, 9, 8, 6, 4, 3, 3]]$['2018-11-15 20:17:31.283019', [27, 13, 12, 11, 7, 3, 2, 1, 0, 0]]$['2018-11-15 20:33:23.904161', [43, 17, 14, 14, 13, 8, 5, 5, 5, 4]]$['2018-11-15 20:49:16.074696', [57, 24, 19, 18, 16, 11, 9, 8, 5, 3]]$['2018-11-15 21:05:08.747100', [33, 28, 15, 13, 12, 12, 11, 5, 3, 2]]$['2018-11-15 21:21:03.459054', [41, 18, 18, 16, 13, 12, 12, 8, 4, 0]]$['2018-11-15 21:36:58.977420', [24, 20, 16, 15, 7, 5, 5, 3, 3, 1]]$['2018-11-15 21:52:53.848454', [30, 26, 20, 18, 17, 10, 10, 7, 3, 2]]$['2018-11-15 22:08:58.699617', [51, 31, 13, 8, 7, 4, 2, 0, 0, 0]]$['2018-11-15 22:24:54.677772', [83, 14, 13, 13, 6, 4, 0, 0, 0, 0]]$['2018-11-15 22:41:18.175756', [17, 17, 8, 7, 6, 2, 0, 0, 0, 0]]$['2018-11-15 22:57:12.399372', [21, 16, 14, 

readSubredditData.py
"""
Module to read and return a readable data unit for each file in SubredditData/

Example Data:
$['2018-11-14 21:35:18.257332', [8, 0, 0, 0, 0, 0, 0, 0, 0, 0]]
"""
from ast import literal_eval

class Reader:
    def __init__(self, fileName : str):
        self.fileName = fileName

    def _read(self):
        #Private Function to read file and return the actual list

        with open("SubredditData/"+self.fileName) as Fobj:
            buffer = Fobj.read()    #File Data

            globalList = [] #Final List

            while (True):
                start = buffer.find('$')
                end = buffer[start+1:].find('$')

                if (end == -1):
                    unit = buffer[start+1:]
                    readyUnit =  literal_eval(unit)
                    globalList.append(readyUnit)
                    break

                unit = buffer[start+1 : end+1]
                readyUnit = literal_eval(unit)
                globalList.append(readyUnit)
                buffer = buffer[end+1:]

        return globalList

    def _HighScore(self, data):
        return max(data)

    def _Mean(self, data):
        return (sum(data)/len(data))

    def _Sigma(self, data):
        return sum(data)

    def getData(self, type = None):
        """
        type:
            'high'  :   returns the highest score in the list (DEFAULT)
            'mean'  :   returns the mean of the values in the list
            'sigma' :   returns the sum of all the values of the list

        Module returns a tuple:
            ([time_1, time_2, ..., time_n], [num_1, num_2, ..., num_n])
                time_i : float
                    [0.00, 0.25, 0.50, ..., 23.75]
                num_i : int
                    [int, int, int, ..., int]
        """

        if (type == None):
            type = 'high'

        if (type not in ['high', 'mean', 'sigma']):
            return -1

        X = []
        y = []

        def typeSelect(data):
            if type == 'high':
                return self._HighScore(data)
            if type == 'mean':
                return self._Mean(data)
            if type == 'sigma':
                return self._Sigma(data)

        rawData = self._read()

        count = 0.00

        X.append(count)
        y.append(typeSelect(rawData[0][1]))
        count += 1

        for i in range(1, len(rawData)):
            timeLower = datetime.strptime(rawData[i-1][0], '%Y-%m-%d %H:%M:%S.%f')
            timeHigher = datetime.strptime(rawData[i][0], '%Y-%m-%d %H:%M:%S.%f')

            timeDiffMins = int( (timeHigher - timeLower).total_seconds() / 60 )

            if (timeDiffMins >= 16):
                missingDataNumber = (timeDiffMins / 15) + 1

                dataLower = rawData[i-1][1]
                dataHigher = rawData[i][1]

                dataMissing = typeSelect(dataLower) + typeSelect(dataHigher)
                dataMissing = dataMissing/2

                for i in range(1, timeDiffMins + 1):
                    X.append(count * 0.25)
                    y.append(dataMissing)
                    count += 1

            else:
                X.append(count * 0.25)
                count += 1

                y.append(typeSelect(rawData[i][1]))

        return (X, y)
